### PR TITLE
Fix 1042 security v2 (review / no-merge)

### PR DESCRIPTION
**Pull request by nathandunn**
This is the initial cut of REMOTE_USER (and other authentication strategies) referred to in #1042.  

Directions forthecoming.

### CODE REVIEW

**Comment by nathandunn** [comment]
Directions for enabling / testing: 


1 - Over-ride in apollo-config.groovy or directly in Config.groovy (for testing):  https://github.com/nathandunn/Apollo/blob/fix_1042_security_v2/grails-app/conf/Config.groovy#L344
2 - to test without wrapping through nginx / Apache, you could uncomment this here to ignore this:  https://github.com/nathandunn/Apollo/blob/fix_1042_security_v2/grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy#L22

**Comment by nathandunn** [comment]
@erasche  If you have time to take a look at this.  I don't have this configured anywhere, but I want to make sure that it would work for Galaxy-land.

**Comment by erasche** [comment]
@nathandunn Heck yes I have time for testing! :D I'm incredible excited for this.

**Comment by erasche** [comment]
I'll look at it tonight.

**Comment by nathandunn** [comment]
Great thanks.  I’m a bit embarrassed about have to pre-register the services: 

https://github.com/nathandunn/Apollo/blob/fix_1042_security_v2/grails-app/services/org/bbop/apollo/PermissionService.groovy#L539 <https://github.com/nathandunn/Apollo/blob/fix_1042_security_v2/grails-app/services/org/bbop/apollo/PermissionService.groovy#L539>

But it seemed to work.   If you have an nginx / apache reference to share (if it fails), I’ll see what I can do on my end, as well.

**Comment by erasche** [comment]
yeah, yeah, on it ;) Let me just get this entrez efetch bug fixed really quickly...

**Comment by nathandunn** [comment]
For whenever, I’ll be out Friday and I’m sure there is more to test.  Was just setting reminders for myself.

**Comment by erasche** [comment]
Okay, after connecting I'm seeing the following, but maybe I'm doing something wrong. I have a dockerized version in https://github.com/erasche/docker-apollo/tree/remote_user (but requires compiled the remote_user_proxy locally... I suppose I could move that to quay.io)

```
apollo_1      | class name remoteUserAuthenticatorService
apollo_1      | 2016-07-07 18:20:48,121 [http-apr-8080-exec-8] ERROR authenticator.RemoteUserAuthenticatorService  - Problem authenticating: org.apache.shiro.authc.AuthenticationException: Authentication failed for token submission [org.apache.shiro.authc.UsernamePasswordToken - admin@local.host, rememberMe=false].  Possible unexpected error? (Typical or expected login exceptions should extend from AuthenticationException).
apollo_1      | class name usernamePasswordAuthenticatorService
apollo_1      | 2016-07-07 18:20:48,121 [http-apr-8080-exec-2] ERROR apollo.PermissionService  - Problem authenticating: org.apache.shiro.authc.AccountException: Null usernames are not allowed by this realm.
apollo_1      | 2016-07-07 18:20:48,125 [http-apr-8080-exec-8] ERROR authenticator.UsernamePasswordAuthenticatorService  - Not implemented without a token
apollo_1      | 2016-07-07 18:20:48,121 [http-apr-8080-exec-3] ERROR apollo.PermissionService  - Problem authenticating: org.apache.shiro.authc.AccountException: Null usernames are not allowed by this realm.
apollo_1      | 2016-07-07 18:20:48,130 [http-apr-8080-exec-8] WARN  apollo.PermissionService  - Failed to authenticate user
apollo_1      | 2016-07-07 18:20:48,130 [http-apr-8080-exec-8] ERROR apollo.UserController  - Failed to authenticate
apollo_1      | 2016-07-07 18:20:48,139 [http-apr-8080-exec-3] ERROR apollo.GroupController  - {error=Cannot cast object 'false' with class 'java.lang.Boolean' to class 'org.codehaus.groovy.grails.web.json.JSONObject'}
apollo_1      | 2016-07-07 18:20:48,139 [http-apr-8080-exec-2] ERROR apollo.UserController  - {error=Cannot cast object 'false' with class 'java.lang.Boolean' to class 'org.codehaus.groovy.grails.web.json.JSONObject'}
```

**Comment by nathandunn** [comment]
Yeah, if you could dockerize both, that would be great, though I'm sure I can work around it.

I'll take a look at that error next week.

**Comment by erasche** [comment]
@nathandunn yeah, sure thing. Will put both on a dockerhub

**Comment by erasche** [comment]
@nathandunn done. `git clone https://github.com/erasche/docker-apollo/; checkout remote_user; docker-compose up`

**Comment by nathandunn** [comment]
Excellent, thanks!  Will take a look next week. 

Nathan

**Comment by nathandunn** [comment]
@erasche  I got this up and running, but not sure how to connect.  

```
git remote -v
origin	https://github.com/erasche/docker-apollo/ (fetch)
origin	https://github.com/erasche/docker-apollo/ (push)
nathandunn:docker-apollo% git merge remote_user
Already up-to-date.
nathandunn:docker-apollo% git status
On branch remote_user
Your branch is up-to-date with 'origin/remote_user'.
nothing to commit, working directory clean
nathandunn:docker-apollo% git pull 
Already up-to-date.
nathandunn:docker-apollo% docker-compose ps
           Name                         Command               State          Ports         
------------------------------------------------------------------------------------------
dockerapollo_apollo_1        catalina.sh run                  Up      8080/tcp             
dockerapollo_chado_1         /docker-entrypoint.sh postgres   Up      5432/tcp             
dockerapollo_db_1            /docker-entrypoint.sh postgres   Up      5432/tcp             
dockerapollo_remote_user_1   dockerize -template /etc/n ...   Up      0.0.0.0:8080->80/tcp 
```

```http://192.168.99.100``` and ```http://192.168.99.100:8080``` fail to connect. 

```http://192.168.99.100:8080/apollo/``` redirects to ```http://192.168.99.100/apollo/``` and then fails.  

I'm going to merge in upstream and try rebuilding.

**Comment by erasche** [comment]
yeah, the redir is broken. Just go to http://192.168.99.100:8080/apollo/annotator/index

**Comment by erasche** [comment]
Sorry about taht

**Comment by nathandunn** [comment]
@erasche  I added a bunch more debugging and I explicitly change the password if authentication fails because the user has set the password somewhere else.  Was thinking how this might with web_services as well.  I think it should be fine so long as you have an internal admin user you can use for just web services.

**Comment by erasche** [comment]
> That worked.  Not sure what the login is, but it shouldn’t matter right?

Yep, you should hit the page and it should figure out that you're who the REMOTE_USER is.

> added a bunch more debugging and I explicitly change the password if authentication fails because the user has set the password somewhere else. Was thinking how this might with web_services as well. I think it should be fine so long as you have an internal admin user you can use for just web services.

Ok, I'll rebuild + retest here soon. The random password is good for web services, sort of API key stand in.

**Comment by nathandunn** [comment]
Okay.  And you are changing the apollo-config.groovy to activate it (looks like you did).  I just fixed something else and retesting.  If the REMOTE_USER is not being passed properly you’ll see a user (hopefully) abcd@124.com <mailto:abcd@124.com>.    Will let you know if the test works.

**Comment by nathandunn** [comment]
K . . it seems to work.  Of course, since the user has access to no organisms it will state that and ask you to log out.   I'm not sure what type of parameters might make sense.  Of course if your default user is already an admin user its probably fine.

**Comment by nathandunn** [comment]
And note that I need to clean things up big-time once it feels like its working.

**Comment by nathandunn** [comment]
K.  Just let me know when to retest.

**Comment by erasche** [comment]
@nathandunn retested, beautiful. Feel free to re-pull images, but the proxying issue won't be fixed, you'll still have to go to /apollo/annotator/index

**Comment by nathandunn** [comment]
👍

**Comment by nathandunn** [comment]
@deepakunni3  Can you review this PR when you have a minute?   I'm going to retest @erasche  docker stuff later today / tomorrow.

**Comment by nathandunn** [comment]
@deepakunni3  Mostly I want to make sure it works for the default case.   I have some doc to write as well.

**Comment by nathandunn** [comment]
@deepakunni3  Doc added.

**Comment by nathandunn** [comment]
@wieckr / @childers  this PR adds REMOTE_USER authentication (via apache / nginx), but could be configured to add additional authenticators.  

This is the [RemoteUserAuthenticatorService](https://github.com/nathandunn/Apollo/blob/2f05e4c88e6605a5cdff5565d90adb7e105aaf4f/grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy) and [UsernamePasswordAuthenticatorService](https://github.com/nathandunn/Apollo/blob/2f05e4c88e6605a5cdff5565d90adb7e105aaf4f/grails-app/services/org/bbop/apollo/authenticator/UsernamePasswordAuthenticatorService.groovy), the default.  

Extending this to use CAS / LDAP shouldn't be too hard if that is what you need.

**Comment by wieckr** [comment]
I would appreciate full LDAP integration.

**Comment by nathandunn** [comment]
@wieckr   We should still be able to do that.   Just wanted to point this out.   If this works, we would add LDAP via the same mechanism.

File: grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    def authenticate(HttpServletRequest request) {
+        User user
+        UsernamePasswordToken authToken = new UsernamePasswordToken()
+        String randomPassword = ClientTokenGenerator.generateRandomString()
+        String passwordHash = new Sha256Hash(randomPassword).toHex()
+        Subject subject
+        try {
+            subject = SecurityUtils.getSubject();
+
+            // TODO: remove .. . . testing only
```
Comment (deepakunni3) [inline grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy]: Was this meant to be removed before this PR is merged?

Comment (nathandunn) [inline grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy]: Definitely.  Thought I'd removed that.  Sorry.

Comment (deepakunni3) [inline grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy]: np!

Comment (nathandunn) [inline grails-app/services/org/bbop/apollo/authenticator/RemoteUserAuthenticatorService.groovy]: I removed it.

**Comment by deepakunni3** [comment]
@nathandunn Tested a normal session with an existing apollo database and a fresh clean apollo database. Both work without any issues.

**Comment by nathandunn** [comment]
Related to #115 WRT to LDAP.  Will post additional comments there.

**Pull request merged**